pallets/msa/src/lib.rs

```diff
@@ -556,6 +571,51 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Retire a MSA
+		///
+		/// ### Events
+		/// - Deposits [`MsaRetired`](Event::MsaRetired) when MSA is retired
+		///
+		/// ### Errors
+		///
+		/// - Returns [`NoKeyExists`](Error::NoKeyExists) if `delegator` does not have an MSA key.
+		/// - Returns [`MoreThanOneKeyExists`](Error::MoreThanOneKeyExists) if the MSA has more than one account key.
+		/// - Returns [`RegisteredProviderCannotBeRetired`](Error::RegisteredProviderCannotBeRetired) if the MSA id is a registered provider
+
+		#[pallet::weight((T::WeightInfo::retire_msa(), DispatchClass::Normal, Pays::No))]
+		pub fn retire_msa(origin: OriginFor<T>) -> DispatchResult {
+			// Check and get the account id from the origin
+			let who = ensure_signed(origin)?;
+
+			// Get the MSA id of the origin
+			let msa_id = Self::get_owner_of(&who).ok_or(Error::<T>::NoKeyExists)?;
+			let delegator = Delegator(msa_id);
+
+			// Dispatches error "NotMsaOwner" if the origin is not the owner of the MSA id.
+			Self::ensure_msa_owner(&who, msa_id)?;
+
+			// Dispatches error "RegisteredProviderCannotBeRetired" if the MSA id is a registered provider
+			ensure!(
+				!Self::is_registered_provider(msa_id),
+				Error::<T>::RegisteredProviderCannotBeRetired,
+			);
+
+			// Dispatches error "MoreThanOneKeyExists" if the MSA has more than one account key.
+			let key_count = Self::get_msa_key_count(msa_id);
+			ensure!(key_count == 1, Error::<T>::MoreThanOneKeyExists);
+
+			// Remove delegator from all delegator<->provider delegations
+			Self::remove_delegator(delegator)?;
```

Comment:
What are your thoughts on maybe calling it `delete_delegator`?

Reply:
I think **remove** is appropriate here because there are other delegations in the map but it could work either way.   It's an internal function so it doesn't matter so much.